src/sparseml/optim/helpers.py

```diff
@@ -207,15 +209,59 @@ def update_recipe_variables(recipe_yaml_str: str, variables: Dict[str, Any]) ->
         # yaml string does not create a dict, return original string
         return recipe_yaml_str
 
-    for key in variables:
-        if key not in container:
-            raise ValueError(
-                f"updating recipe variable {key} but {key} is not currently "
-                "set in existing recipe. Set the variable in the recipe in order "
-                "to overwrite it."
-            )
+    key_found = False
+
+    if check_if_staged_recipe(container):
+        for var_key in variables.keys():
+            for container_key, container_value in container.items():
+
+                if isinstance(container_value, dict):
+                    # checking contents of a stage
+                    for stage_key, stage_value in container_value.items():
+
+                        # checking contents of the modifiers list
+                        if isinstance(stage_value, list):
+                            for i, modifier in enumerate(stage_value):
+                                if var_key in modifier.keys():
+                                    container[container_key][stage_key][i][
+                                        var_key
+                                    ] = variables[var_key]
+                                    key_found = True
+
+                        # checking stage variables
+                        else:
+                            if var_key == stage_key:
+                                container[container_key][var_key] = variables[var_key]
+                                key_found = True
+
+                # checking contents of global variables
+                else:
+                    if var_key == container_key:
+                        container[var_key] = variables[var_key]
+                        key_found = True
+
+    else:
+        for var_key in variables.keys():
+            for key, value in container.items():
+                # checking contents of the modifiers list
+                if isinstance(value, list):
+                    for i, modifier in enumerate(value):
+                        if var_key in modifier.keys():
+                            container[key][i][var_key] = variables[var_key]
+                            key_found = True
+
+                else:
+                    if var_key == key:
+                        container[var_key] = variables[var_key]
+                        key_found = True
+
+    if not key_found:
```

Comment:
this needs to be moved inside the loop, if any key is missing we want to raise (also notice that `var_key` is a loop variable, so we would only print the last variable regardless if that one was found or not

Reply:
Excellent catch, thank you Ben!